crates/ra_hir/src/nameres/per_ns.rs

```diff
@@ -1,78 +1,93 @@
-        PerNs { types: None, values: None }
+        PerNs { types: None, values: None, macros: None }
     }
 }
 
 impl<T> PerNs<T> {
     pub fn none() -> PerNs<T> {
-        PerNs { types: None, values: None }
+        PerNs { types: None, values: None, macros: None }
     }
 
     pub fn values(t: T) -> PerNs<T> {
-        PerNs { types: None, values: Some(t) }
+        PerNs { types: None, values: Some(t), macros: None }
     }
 
     pub fn types(t: T) -> PerNs<T> {
-        PerNs { types: Some(t), values: None }
+        PerNs { types: Some(t), values: None, macros: None }
     }
 
     pub fn both(types: T, values: T) -> PerNs<T> {
-        PerNs { types: Some(types), values: Some(values) }
+        PerNs { types: Some(types), values: Some(values), macros: None }
     }
 
-    pub fn is_none(&self) -> bool {
-        self.types.is_none() && self.values.is_none()
+    pub fn macros(macro_: MacroDef) -> PerNs<T> {
+        PerNs { types: None, values: None, macros: Some(macro_) }
     }
 
-    pub fn is_both(&self) -> bool {
-        self.types.is_some() && self.values.is_some()
+    pub fn is_none(&self) -> bool {
+        self.types.is_none() && self.values.is_none() && self.macros.is_none()
     }
 
-    pub fn take(self, namespace: Namespace) -> Option<T> {
-        match namespace {
-            Namespace::Types => self.types,
-            Namespace::Values => self.values,
-        }
+    pub fn is_all(&self) -> bool {
+        self.types.is_some() && self.values.is_some() && self.macros.is_some()
     }
 
     pub fn take_types(self) -> Option<T> {
-        self.take(Namespace::Types)
+        self.types
     }
 
     pub fn take_values(self) -> Option<T> {
-        self.take(Namespace::Values)
+        self.values
     }
 
-    pub fn get(&self, namespace: Namespace) -> Option<&T> {
-        self.as_ref().take(namespace)
+    pub fn get_macros(&self) -> Option<MacroDef> {
+        self.macros
+    }
+
+    pub fn only_macros(&self) -> PerNs<T> {
+        PerNs { types: None, values: None, macros: self.macros }
     }
 
     pub fn as_ref(&self) -> PerNs<&T> {
-        PerNs { types: self.types.as_ref(), values: self.values.as_ref() }
+        PerNs { types: self.types.as_ref(), values: self.values.as_ref(), macros: self.macros }
     }
 
     pub fn or(self, other: PerNs<T>) -> PerNs<T> {
-        PerNs { types: self.types.or(other.types), values: self.values.or(other.values) }
+        PerNs {
+            types: self.types.or(other.types),
+            values: self.values.or(other.values),
+            macros: self.macros.or(other.macros),
+        }
     }
 
     pub fn and_then<U>(self, f: impl Fn(T) -> Option<U>) -> PerNs<U> {
-        PerNs { types: self.types.and_then(&f), values: self.values.and_then(&f) }
+        PerNs {
+            types: self.types.and_then(&f),
+            values: self.values.and_then(&f),
+            macros: self.macros,
+        }
     }
 
     pub fn map<U>(self, f: impl Fn(T) -> U) -> PerNs<U> {
```

Comment:
I feel like `and_then` and `map` should transform macros as well. Does this break anything?

Reply:
They have different types. I don't know how to do it better. `and_then` is not used. `map` is only called with `map(Resolution::Def)`. Actually, there are only two types, `PerNs<Resolution>` and `PerNs<ModuleDef>`.

Reply:
Ah, righ, `PerNs` is not generic on macro ns... Yeah, this sort-of makes sense, but at the same time is pretty ugly.... I don't know a better solution though, and existing `Either` stuff is not really pretty anyway, so I am fine with it!

Reply:
Maybe just rename to `map_type_value()` or something? `and_then()` can be simply removed.

Reply:
Let's remove and_then, and leave `map` as is, for know. I have a feeling that we should remove `.map` as well. Like, we use map to construct `PerNs<Resolution>` in resolve, but this type doesn't make sense